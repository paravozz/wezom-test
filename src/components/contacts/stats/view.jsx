import React from 'react';
import { Typography, Col, Row, Statistic, Descriptions } from 'antd';
import {
	countGenders,
	getGenderPredominateMessage,
	countNationalities,
	CountryData,
} from 'utils';


const View = (props) => {
	const { contacts, loading } = props;

	if (!contacts?.length || loading) {
		return (
			<Row className="box" />
		);
	}

	function renderNationalitiesStats(_contacts) {
		const natCount = countNationalities(_contacts);

		return (
			<Descriptions title="Nationalities">
				{Object.entries(natCount).map(([key, value]) => (
					<Descriptions.Item key={key} label={CountryData[key].name}>
						{value} contacts
					</Descriptions.Item>
				))}
			</Descriptions>
		);
	}

	const genderStats = countGenders(contacts);
	const genderPredominateMessage = getGenderPredominateMessage(genderStats);

	return (
		<div className="box">
			<Typography.Title level={3}>Statistic</Typography.Title>
			<Row gutter={24}>
				<Col className="_flex-noshrink">
					<Statistic title="Collection size" value={contacts.length} />
				</Col>
				<Col className="_flex-noshrink">
					<Row gutter={12}>
						<Col span={6}>
							<Statistic title="Males" value={genderStats.male} />
						</Col>
						<Col span={6}>
							<Statistic title="Females" value={genderStats.female} />
						</Col>
						<Col span={6}>
							<Statistic title="Indeterminate" value={genderStats.indeterminate} />
						</Col>
						<Col span={24}>
							<Typography.Text mark>
								{genderPredominateMessage}
							</Typography.Text>
						</Col>
					</Row>
				</Col>
			</Row>
			<br />
			<Row>
				{renderNationalitiesStats(contacts)}
			</Row>
		</div>
	);
};


export { View };
