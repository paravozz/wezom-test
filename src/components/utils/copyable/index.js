export { default as Copyable } from './view';
