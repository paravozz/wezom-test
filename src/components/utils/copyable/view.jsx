import React from 'react';

import { CopyToClipboard } from 'react-copy-to-clipboard';
import { CopyOutlined, CheckOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';

import './style.scss';


export default ({ children, text }) => {
	const [isCopied, setIsCopied] = React.useState(false);


	React.useEffect(() => {
		if (isCopied) {
			setTimeout(() => {
				setIsCopied(false);
			}, 3000);
		}
	}, [isCopied]);


	return (
		<div className="copyable--container">
			<CopyToClipboard onCopy={() => setIsCopied(true)} text={text}>
				<Tooltip title={isCopied ? 'Copied' : 'Copy'}>
					{
						isCopied
							? <CheckOutlined style={{ color: '#52c41a' }} />
							: <CopyOutlined style={{ color: '#1890ff' }} />
					}
				</Tooltip>
			</CopyToClipboard>
			{children}
		</div>
	)
};
