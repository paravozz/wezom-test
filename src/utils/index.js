export {
	provide,
	provideWithRandomError,
} from './provide';
export {
	randomIntegerInRange,
} from './random';
export { redirect } from './redirect';
export {
	getFromStorage,
	setToStorage,
} from './local-storage-api';
export {
	countGenders,
	getGenderPredominateMessage,
	countNationalities,
} from './contacts-stats';
export { CountryData } from './countries';
export {
	getFullName,
	getGenderName,
	mapGenderToItsName,
	paginateCollection
} from './contacts';
