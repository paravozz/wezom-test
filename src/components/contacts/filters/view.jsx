import React from 'react';
import { Row, Input, Select, Col, Checkbox, Button } from 'antd';
import { CloseOutlined } from '@ant-design/icons';
import { CountryData } from 'utils';
import { getGenderName } from 'utils';


const View = (props) => {
	const {
		loading,
		filters,
		filterContacts,
	} = props;

	return (
		<div className="box">
			<Row align="middle">
				<Col className="_flex-grow">
					<Row gutter={[6, 6]} align="middle">
						<Col xs={24} lg={11}>
							<Row>
								<Input.Search
									loading={loading}
									placeholder="Search by full name"
									size="large"
									onChange={e => filterContacts({ ...filters, name: e.target.value })}
									value={filters.name}
								/>
							</Row>
						</Col>
						<Col xs={24} sm={10} lg={4}>
							<Row>
								<Select
									id="genderSelect"
									loading={loading}
									style={{ width: '100%' }}
									allowClear
									placeholder="Gender"
									size="large"
									onChange={value => filterContacts({ ...filters, gender: value })}
									value={filters.gender || undefined}
								>
									{['male', 'female', 'indeterminate'].map(k => (
										<Select.Option key={k} value={k}>
											{getGenderName(k)}
										</Select.Option>
									))}
								</Select>
							</Row>
						</Col>
						<Col xs={24} sm={14} lg={5}>
							<Row>
								<Select
									id="natSelect"
									loading={loading}
									style={{ width: '100%' }}
									mode="multiple"
									maxTagCount={2}
									allowClear
									size="large"
									placeholder="Nationality"
									onChange={value => filterContacts({ ...filters, nat: value })}
									value={filters.nat}
								>
									{Object.entries(CountryData).map(([key, data]) => (
										<Select.Option key={key} value={key}>{data.name}</Select.Option>
									))}
								</Select>
							</Row>
						</Col>
						<Col xs={24} sm={6} lg={4}>
							<Row>
								<Checkbox
									disabled={loading}
									onChange={(e) => filterContacts({ ...filters, author: e.target.checked })}
								>
									I am creator
								</Checkbox>
							</Row>
						</Col>
					</Row>
				</Col>
				<Col className="_flex-noshrink">
					<Row>
						<Button
							disabled={loading}
							onClick={() => filterContacts(null)}
							type="link"
						>
							<CloseOutlined /> Clear
						</Button>
					</Row>
				</Col>
			</Row>
		</div>
	);
};


export { View };
