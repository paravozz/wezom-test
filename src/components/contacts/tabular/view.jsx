import React from 'react';
import { Table, Typography, Tag, Avatar, Empty } from 'antd';
import { UserOutlined } from '@ant-design/icons';
import moment from 'moment';
import { CountryData, getFullName } from 'utils';
import { Copyable } from 'components/utils';


const View = (props) => {
	const { contacts, loading } = props;

	if (!contacts) {
		return (
			<>
				<Empty image={Empty.PRESENTED_IMAGE_SIMPLE} />
			</>
		)
	}

	function renderFullName({ title, first, last }) {
		return getFullName(title, first, last);
	}

	function renderAvatar(pictureUrl) {
		return (
			<Avatar size={40} icon={<UserOutlined />} src={pictureUrl} />
		);
	}

	function renderBirthday({ date, age }) {
		return (
			<div>
				{moment(date).format('dddd, M/D/YYYY, H:m:s A')}
				<br />
				{age} years
			</div>
		);
	}

	function renderEmail(email) {
		return (
			<Copyable text={email}>
				<Typography.Link ellipsis href={`mailto:${email}`}>
					{email}
				</Typography.Link>
			</Copyable>
		);
	}

	function renderPhone(phone) {
		return (
			<Copyable text={phone}>
				<Typography.Link ellipsis href={`tel:${phone}`}>
					{phone}
				</Typography.Link>
			</Copyable>
		);
	}

	function renderLocation({ country, street, city, state, postcode }) {
		return (
			<Copyable
				text={(
					`[${country}] `
					+ `${street.number} ${street.name}, `
					+ `${city}, ${state} ${postcode}`
				)}
			>
				<Typography.Paragraph ellipsis={{ rows: 3 }}>
					<strong>/{country}/</strong>
					<br />
					{(
						`${street.number}, ${street.name}`
						+ `${city}, ${state} ${postcode}`
					)}
				</Typography.Paragraph>
			</Copyable>
		);
	}

	function renderNationality(nat) {
		return (
			<Tag color={CountryData[nat].color}>{CountryData[nat].name}</Tag>
		)
	}

	const dataSource = (
		contacts
			.map((contact, index) => ({
				key: index,
				avatar: renderAvatar(contact.picture.thumbnail),
				fullName: renderFullName(contact.name),
				birthday: renderBirthday(contact.dob),
				email: renderEmail(contact.email),
				phone: renderPhone(contact.phone),
				location: renderLocation(contact.location),
				nationality: renderNationality(contact.nat),
			}))
	);

	const columns = [
		{
			title: 'Avatar',
			dataIndex: 'avatar',
			key: 'avatar',
			fixed: 'left',
			width: 80,
		},
		{
			title: 'Full name',
			dataIndex: 'fullName',
			key: 'fullName',
			width: 155,
			sorter: (a, b) => a.fullName.length - b.fullName.length,
			sortDirections: ['descend', 'ascend'],
		},
		{
			title: 'Birthday',
			dataIndex: 'birthday',
			key: 'birthday',
			width: 250,
		},
		{
			title: 'Email',
			dataIndex: 'email',
			key: 'email',
			width: 155,
		},
		{
			title: 'Phone',
			dataIndex: 'phone',
			key: 'phone',
			width: 155,
		},
		{
			title: 'Location',
			dataIndex: 'location',
			key: 'location',
			width: 250,
		},
		{
			title: 'Nationality',
			dataIndex: 'nationality',
			key: 'nationality',
			width: 155,
			align: 'right',
		},
	];


	return (
		<Table
			dataSource={dataSource}
			columns={columns}
			scroll={{ x: '100vw' }}
			loading={loading}
			pagination={{ position: ['bottomCenter'] }}
		/>
	);
};


export { View };
