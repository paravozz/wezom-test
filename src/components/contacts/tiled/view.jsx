import React from 'react';
import { Typography, Tag, Row, Col, Card, Divider, Empty, Spin } from 'antd';
import { CountryData, getFullName, paginateCollection } from 'utils';
import { Copyable } from 'components/utils';


function renderFullName({ title, first, last }) {
	return getFullName(title, first, last);
}

function renderAvatar(pictureUrl) {
	return (
		<img src={pictureUrl} alt="contact-cover" />
	);
}

function renderBirthday({ age }) {
	return `${age} years`
}

function renderEmail(email) {
	return (
		<Copyable text={email}>
			<Typography.Link ellipsis href={`mailto:${email}`}>
				{email}
			</Typography.Link>
		</Copyable>
	);
}

function renderPhone(phone) {
	return (
		<Copyable text={phone}>
			<Typography.Link ellipsis href={`tel:${phone}`}>
				{phone}
			</Typography.Link>
		</Copyable>
	);
}

function renderLocation({ country, street, city, state, postcode }) {
	return (
		<Copyable
			text={(
				`[${country}] `
				+ `${street.number} ${street.name}, `
				+ `${city}, ${state} ${postcode}`
			)}
		>
			<Typography.Paragraph ellipsis={{ rows: 3 }}>
				<strong>/{country}/</strong>
				<br />
				{(
					`${street.number}, ${street.name}`
					+ `${city}, ${state} ${postcode}`
				)}
			</Typography.Paragraph>
		</Copyable>
	);
}

function renderNationality(nat) {
	return (
		<Tag color={CountryData[nat].color}>{CountryData[nat].name}</Tag>
	)
}

export function transformContacts(pagination, contacts) {
	let baseCollection = (
		pagination
			? paginateCollection(contacts, pagination.offset, pagination.limit)
			: contacts
	);

	return baseCollection.map((contact, index) => ({
		key: index,
		avatar: renderAvatar(contact.picture.large),
		fullName: renderFullName(contact.name),
		birthday: renderBirthday(contact.dob),
		email: renderEmail(contact.email),
		phone: renderPhone(contact.phone),
		location: renderLocation(contact.location),
		nationality: renderNationality(contact.nat),
	}));
}


const View = (props) => {
	const { contacts, loading, pagination } = props;

	if (!contacts?.length || loading) {
		return (
			<>
				{loading && <Spin />}
				<Empty image={Empty.PRESENTED_IMAGE_SIMPLE} />
			</>
		)
	}

	return (
		<Row gutter={[8, 8]}>
			{transformContacts(pagination, contacts).map(contact => (
				<div key={contact.key} style={{ width: '33.3333%', maxWidth: '33.3333%' }}>
					<Col>
						<Card
							hoverable
							cover={contact.avatar}
						>
							<Card.Meta
								title={(
									<Typography.Text>
										<Typography.Text>{contact.fullName}</Typography.Text>
										&nbsp;
										<Typography.Text type="secondary">
											({contact.birthday})
										</Typography.Text>
									</Typography.Text>
								)}
								description={(
									<>
										{contact.email}
										{contact.phone}
										{contact.location}
										<Divider/>
										{contact.nationality}
									</>
								)}
							/>
						</Card>
					</Col>
				</div>
			))}
		</Row>
	);
};


export { View };
