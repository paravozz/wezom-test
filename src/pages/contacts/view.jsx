import React from 'react';

import { Radio, Row, Col, Typography, Pagination, Tooltip, Button } from 'antd';
import { DatabaseOutlined, AppstoreOutlined, ReloadOutlined } from '@ant-design/icons';

import { getFromStorage, randomIntegerInRange, setToStorage } from 'utils';
import {
	TabularContacts,
	StatsForContacts,
	FiltersForContacts,
} from 'components';
import { TiledContacts } from '../../components/contacts/tiled';


const __VIEW_MODE_KEY__ = 'app__view_mode';


const View = (props) => {
	const {
		totalCount,
		fetchContacts,
		pagination,
		setPagination,
		contacts,
	} = props;

	const [viewMode, setViewMode] = React.useState(getFromStorage(__VIEW_MODE_KEY__) || 'tabular');

	function fetchContactsRemote() {
		fetchContacts(totalCount || randomIntegerInRange(1, 1000));
	}


	React.useEffect(() => {
		fetchContactsRemote();
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	React.useEffect(() => {
		setToStorage(__VIEW_MODE_KEY__, viewMode);
	}, [viewMode]);


	const radioOptions = [
		{ label: <DatabaseOutlined />, value: 'tabular' },
		{ label: <AppstoreOutlined />, value: 'tiled' },
	];

	const componentForViewMode = {
		tiled: (
			<>
				<TiledContacts />
				<Pagination
					onChange={(page, pageSize) => (
						setPagination({
							offset: pageSize * (page - 1),
							limit: pageSize,
						})
					)}
					current={pagination.offset / pagination.limit + 1}
					pageSize={pagination.limit}
					pageSizeOptions={[6, 12, 24, 36, 48]}
					total={contacts.length}
				/>
			</>
		),
		tabular: <TabularContacts />,
	};


	return (
		<div className={'page page--contacts'}>
			<Row justify="space-between">
				<Typography.Title level={2}>Contacts</Typography.Title>
				<Col>
					<Tooltip title="Reload data">
				  		<Button
							type="dashed"
							shape="circle"
							icon={<ReloadOutlined />}
							onClick={() => fetchContactsRemote()}
						/>
					</Tooltip>
					<Radio.Group
						options={radioOptions}
						onChange={(e) => setViewMode(e.target.value)}
						value={viewMode}
						optionType="button"
						buttonStyle="solid"
					/>
				</Col>
			</Row>
			<FiltersForContacts />
			{componentForViewMode[viewMode]}
			<StatsForContacts />
		</div>
	);
};


export { View };
