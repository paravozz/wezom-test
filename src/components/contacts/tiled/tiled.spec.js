/* eslint-disable */
import React from 'react';
import { View as TiledContacts } from './view.jsx';
import { Empty, Card, Spin } from 'antd';
import { paginateCollection } from 'utils';


jest.mock('utils');
const setUp = (props) => shallow(<TiledContacts {...props} />);
const setUpRendered = (props) => render(<TiledContacts {...props} />);

describe('TiledContacts', () => {
	let component;

	beforeEach(() => {
		jest.resetModules();
		component = setUp();
	});

	it('should render without crashing', () => {
		component = setUpRendered({ contacts });
		expect(component).toMatchSnapshot();
	});

	it('should render without redux props', () => {
		expect(component)
	});

	it('should contain Empty', () => {
		expect((component.find(Empty))?.length).toBe(1);
	});

	it('should contain Spin if loading is true', () => {
		component = setUp({ loading: true });
		expect((component.find(Spin))?.length).toBe(1);
	});

	it('should render contacts', () => {
		component = setUp({ contacts });

		expect((component.find(Empty))?.length).toBe(0);
		expect((component.find(Card))?.length).toBe(contacts.length);
	});

	it('should paginate collection if pagination prop is provided', () => {
		const pagination = { limit: 1, offset: 1 };
		paginateCollection.mockReturnValue([contacts[1]]);

		component = setUp({ contacts, pagination });

		expect(paginateCollection).toBeCalled();
		expect(paginateCollection).toBeCalledWith(contacts, pagination.offset, pagination.limit);
	});

	it('should return original collection if no pagination prop is provided', () => {
		component = setUp({ contacts });

		expect(paginateCollection).not.toBeCalled();
	});
});
