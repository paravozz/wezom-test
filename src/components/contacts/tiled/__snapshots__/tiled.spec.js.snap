// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TiledContacts should render without crashing 1`] = `
<div
  class="ant-row"
  style="margin-left:-4px;margin-right:-4px;margin-top:-4px;margin-bottom:4px"
>
  <div
    style="width:33.3333%;max-width:33.3333%"
  >
    <div
      class="ant-col"
      style="padding-left:4px;padding-right:4px;padding-top:4px;padding-bottom:4px"
    >
      <div
        class="ant-card ant-card-bordered ant-card-hoverable"
      >
        <div
          class="ant-card-cover"
        >
          <img
            alt="contact-cover"
            src="https://randomuser.me/api/portraits/women/42.jpg"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                <span
                  class="ant-typography"
                >
                  <span
                    class="ant-typography"
                  />
                   
                  <span
                    class="ant-typography ant-typography-secondary"
                  >
                    (43 years)
                  </span>
                </span>
              </div>
              <div
                class="ant-card-meta-description"
              >
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="mailto:hazel.chen@example.com"
                  >
                    hazel.chen@example.com
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="tel:(779)-242-0785"
                  >
                    (779)-242-0785
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <div
                    class="ant-typography ant-typography-ellipsis"
                  >
                    <strong>
                      /New Zealand/
                    </strong>
                    <br />
                    6257, Esmonde RoadTimaru, Wellington 27976
                  </div>
                </div>
                <div
                  class="ant-divider ant-divider-horizontal"
                  role="separator"
                />
                <span
                  class="ant-tag ant-tag-has-color"
                  style="background-color:#9e1068"
                >
                  New Zealander
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style="width:33.3333%;max-width:33.3333%"
  >
    <div
      class="ant-col"
      style="padding-left:4px;padding-right:4px;padding-top:4px;padding-bottom:4px"
    >
      <div
        class="ant-card ant-card-bordered ant-card-hoverable"
      >
        <div
          class="ant-card-cover"
        >
          <img
            alt="contact-cover"
            src="https://randomuser.me/api/portraits/men/7.jpg"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                <span
                  class="ant-typography"
                >
                  <span
                    class="ant-typography"
                  />
                   
                  <span
                    class="ant-typography ant-typography-secondary"
                  >
                    (62 years)
                  </span>
                </span>
              </div>
              <div
                class="ant-card-meta-description"
              >
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="mailto:troy.lucas@example.com"
                  >
                    troy.lucas@example.com
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="tel:07-3482-4477"
                  >
                    07-3482-4477
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <div
                    class="ant-typography ant-typography-ellipsis"
                  >
                    <strong>
                      /Australia/
                    </strong>
                    <br />
                    5314, Poplar DrHobart, South Australia 2836
                  </div>
                </div>
                <div
                  class="ant-divider ant-divider-horizontal"
                  role="separator"
                />
                <span
                  class="ant-tag ant-tag-magenta"
                >
                  Australian
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style="width:33.3333%;max-width:33.3333%"
  >
    <div
      class="ant-col"
      style="padding-left:4px;padding-right:4px;padding-top:4px;padding-bottom:4px"
    >
      <div
        class="ant-card ant-card-bordered ant-card-hoverable"
      >
        <div
          class="ant-card-cover"
        >
          <img
            alt="contact-cover"
            src="https://randomuser.me/api/portraits/women/86.jpg"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                <span
                  class="ant-typography"
                >
                  <span
                    class="ant-typography"
                  />
                   
                  <span
                    class="ant-typography ant-typography-secondary"
                  >
                    (52 years)
                  </span>
                </span>
              </div>
              <div
                class="ant-card-meta-description"
              >
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="mailto:angela.henry@example.com"
                  >
                    angela.henry@example.com
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <a
                    class="ant-typography ant-typography-ellipsis"
                    href="tel:0131 413 2855"
                  >
                    0131 413 2855
                  </a>
                </div>
                <div
                  class="copyable--container"
                >
                  <span
                    aria-label="copy"
                    class="anticon anticon-copy"
                    role="img"
                    style="color:#1890ff"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="copy"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                      />
                    </svg>
                  </span>
                  <div
                    class="ant-typography ant-typography-ellipsis"
                  >
                    <strong>
                      /United Kingdom/
                    </strong>
                    <br />
                    6473, Park LaneYork, West Midlands M86 0DJ
                  </div>
                </div>
                <div
                  class="ant-divider ant-divider-horizontal"
                  role="separator"
                />
                <span
                  class="ant-tag ant-tag-geekblue"
                >
                  British
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
